src/Storages/WindowView/StorageWindowView.cpp

```diff
@@ -0,0 +1,1382 @@
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+}
+
+Block & StorageWindowView::getHeader() const
+{
+    std::lock_guard lock(sample_block_lock);
+    if (!sample_block)
+    {
+        sample_block = InterpreterSelectQuery(
+            getFinalQuery(), window_view_context, getParentStorage(), nullptr,
+            SelectQueryOptions(QueryProcessingStage::Complete)).getSampleBlock();
+
+        for (size_t i = 0; i < sample_block.columns(); ++i)
+            sample_block.safeGetByPosition(i).column = sample_block.safeGetByPosition(i).column->convertToFullColumnIfConst();
+    }
+    return sample_block;
+}
+
+StoragePtr StorageWindowView::getParentStorage() const
+{
+    if (!parent_storage)
+        parent_storage = DatabaseCatalog::instance().getTable(select_table_id, getContext());
+    return parent_storage;
+}
+
+StoragePtr StorageWindowView::getInnerStorage() const
+{
+    if (!inner_storage)
+        inner_storage = DatabaseCatalog::instance().getTable(inner_table_id, getContext());
+    return inner_storage;
+}
+
+ASTPtr StorageWindowView::getFetchColumnQuery(UInt32 w_start, UInt32 w_end) const
+{
+    auto res_query = std::make_shared<ASTSelectQuery>();
+    auto select = std::make_shared<ASTExpressionList>();
+    select->children.push_back(std::make_shared<ASTAsterisk>());
+    res_query->setExpression(ASTSelectQuery::Expression::SELECT, select);
+    res_query->setExpression(ASTSelectQuery::Expression::TABLES, std::make_shared<ASTTablesInSelectQuery>());
+    auto tables_elem = std::make_shared<ASTTablesInSelectQueryElement>();
+    auto table_expr = std::make_shared<ASTTableExpression>();
+    res_query->tables()->children.push_back(tables_elem);
+    tables_elem->table_expression = table_expr;
+    tables_elem->children.push_back(table_expr);
+    table_expr->database_and_table_name = std::make_shared<ASTTableIdentifier>(inner_table_id);
+    table_expr->children.push_back(table_expr->database_and_table_name);
+
+    if (is_tumble)
+    {
+        /// SELECT * FROM inner_table PREWHERE window_id_name == w_end
+        /// (because we fire at the end of windows)
+        auto func_equals = makeASTFunction("equals", std::make_shared<ASTIdentifier>(window_id_name), std::make_shared<ASTLiteral>(w_end));
+        res_query->setExpression(ASTSelectQuery::Expression::PREWHERE, func_equals);
+    }
+    else
+    {
+        /// SELECT * FROM inner_table PREWHERE window_id_name in if( w_start < w_end, [w_end, w_end - window_size], [])
+        auto func_array = makeASTFunction("array");
+        while (w_start < w_end)
+        {
+            /// slice_num_units = std::gcd(hop_num_units, window_num_units);
+            /// TODO: why gcd?
+            func_array ->arguments->children.push_back(std::make_shared<ASTLiteral>(w_end));
+            w_end = addTime(w_end, window_kind, -1 * slice_num_units, *time_zone);
```

Comment:
Did not understand why as `slice_num_units` need to use `std::gcd(hop_num_units, window_num_units)` here?

Reply:
The hopping windows might be overlapped. For a hopping window with  window_size=3 slice=1, the windows might be [1,3],[2,4],[3,5], which will cause recomputation. We use `std::gcd(hop_num_units, window_num_units)` as the new window size to split the overlapped windows into non-overlapped, in this case , the `slice_num_units` will be `gcd(1,3)=1' and the overlapped windows will split into  [1], [2], [3]... We compute each split window into mergeable state and merge them when the window is triggering.

Reply:
btw looks like `HOT/TUMBLE_START/END` functions are not quite deterministic as `01049_window_view_window_functions` test is flacky and fails from time to time (I didn't change anything in implementation of `HOT/TUMBLE_START/END` functions, but in CI this test randomly either successfully finishes or fails, for example see `Stateless tests flaky check`). 

Also may be for the same reason some other tests fail also randomly, because when I run them locally they pass, but for example in [this check](https://clickhouse-test-reports.s3.yandex.net/8331/7fae3b1cc9c0eb9dcce91117db918ed81d347c30/fast_test.html#fail1) all tests apart from 2 passed, but in [this check](https://clickhouse-test-reports.s3.yandex.net/8331/7fae3b1cc9c0eb9dcce91117db918ed81d347c30/functional_stateless_tests_(release).html#fail1) (also release build without any sanitizers) those which passed in the previous check failed now.

Haven't yet found how to fix this properly, any idea?

Reply:
`01049_window_view_window_functions` might failed by different timezone. The test

```
SELECT toDateTime(TUMBLE_START(toDateTime('2020-01-09 12:00:01'), INTERVAL '1' DAY, 'US/Samoa'), 'US/Samoa');
```
on my machine( Asia/Shanghai) will result in 
```
 2020-01-08 00:00:00
```

But if I add a timezone to the test.
```
SELECT toDateTime(TUMBLE_START(toDateTime('2020-01-09 12:00:01','US/Samoa'), INTERVAL '1' DAY, 'US/Samoa'), 'US/Samoa');
```
the result will be 
```
2020-01-09 00:00:00
```

Reply:
oh, yes, looks so, thanks

Reply:
hm, I feel like I added timezone everywhere and updated .reference, but still different results:
```
2021-11-26 10:21:52  SELECT TUMBLE_END(toDateTime('2020-01-09 12:00:01', 'US/Samoa'), INTERVAL '1' DAY, 'US/Samoa');
2021-11-26 10:21:52 -2020-01-10 14:00:00
2021-11-26 10:21:52 +2020-01-10 08:00:00
```

Reply:
It is caused because `timezone` is not specified in the return type of function `TUMBLE_START/END...`. I have pushed a new commit to fix this.

Reply:
Ok, thanks!